(feat): Add an interface for external vendors to contribute the UI (#9502)
openshift/console: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/odf-external-providers/README.md

## Contributing to the `CreateStorageSystem` Wizard:

OpenShift Data Foundation (ODF) operator supports connecting to an external vendor storage system.

All external vendors can contribute their UI via the `Create storage class` step of `CreateStorageSystem` wizard.

Any new external storage vendor would require to add a new entry to  [`ODF_EXTERNAL_PROVIDERS`](https://github.com/openshift/console/blob/master/frontend/packages/ceph-storage-plugin/src/odf-external-providers/external-providers.ts) with the following details:

```JS
{
    /* The product name of the external system to be displayed in the dropdown selection */
    label: string;

    /* The model which refers to the sub-system resource. */
    model: ExtensionK8sModel;

    /* 
     * id is a unique identifier that will be used in naming StorageSystems

>      * and sub-StorageSystems e.g `odf-<name>-storage-system`.

Reviewer: ```suggestion
     * and sub-StorageSystems e.g `odf-<id>-storage-system`.
```

     */
    id: string;

    /* A React Functional Component to render the connection details of the external provider.*/
    Component: React.FunctionComponent<ComponentProps>;

    /* Handler function to create request payload for the external sub-StorageSystem */
    createPayload: CreatePayload;
};

```

### `label`

> The product name of the external system to be displayed in the dropdown selection of wizard.

Reviewer: ```suggestion
The product name of the external provider to be displayed in the dropdown selection of wizard.
```

### `model`

> This is a reference to external vendor CRD. The model should be defined as:

Reviewer: ```suggestion
This is a reference to the underlying CRD that is managed by ODF Storage System. The model should be defined as:
```
Vendors can have multiple CRDs.

* A new folder should be created for a provider as `./odf-external-provider/{id-of-provider}`. The name of the folder should be the same as `id`. [See`id`](#id).
* Define your Component inside the newly created folder by adding a file with the same name as of the folder.
* The Component must use the [Patternfly's form components](https://www.patternfly.org/v4/components/form) e.g `FormGroup`, `TextInput`, `Radio` to make the design consistent with the rest of the wizard and OpenShift console.
* All form components should be controlled forms i.e using `onChange` event handler to control the value of the input elements.
* The component should externalize the strings for i18n by namespace: `ceph-storage-plugin~`

The Component has following three props:

##### 1. `form`
This is the state of the form that will be rendered by the Component. The state is a plain javascript object where each `key` represents a unique name of the form field and `value` represents the value of the field.

For instance,

```js
   //  Intial render state

>    state = {};
> 
>    // Post render state
>    state = {
>        username: 'kubeadmin',
>        password: 'gdjshdjshdX==',
>    }

Reviewer: ```suggestion
   state: FormState = {};

   // Post render state
   state = {
       username: 'kubeadmin',
       password: 'gdjshdjshdX==',
   }
```
Vendors should be able to leverage typescript. Make use of generics to provide vendors with support for typescript.

  For instance:

  ```js
    <FormGroup label={t('ceph-storage-plugin~Username')} ... >
        <TextInput ... onChange={(value => setState('username', value)}/> 
    </FormGroup>
  ```

  b.  `value`: This is a string which represents the actual value for the form field.

#### 3. `t`
It is the translation function for i18n and can be used as follows:
```js
 <FormGroup label={t('ceph-storage-plugin~Username')} ... >
    ...
</FormGroup>

```
### `createPayload`

This is a handler for creating the request payload for the external system.
Define this component in the same file which hosts your component.
It takes two arguments:

  1. `storageClass`: name of the storage class. It is a string value.
  2. `form`: the state of the form. [See more](#1-form)
  3. `model`:  a reference to external vendor CRD. [See more](#model)
  4. `name`: the name of the external vendor system.

## Example:

Sample file url: `/src/components/create-storage-system/odf-external-providers/ibm-flash/ibm-flash.tsx`

```js
import * as React from 'react';
import { Form, FormGroup, TextInput } from '@patternfly/react-core';

import { ComponentProps, CreatePayload } from '../types';

> export const CreateFlashSystem: React.FC<ComponentProps> = ({ form, setForm, t }) => {

Author: Passing that so it cant be missed.

> export const CreateFlashSystem: React.FC<ComponentProps> = ({ form, setForm, t }) => {

Reviewer: Not really. You cannot force users to use a prop. So I am not sure how we are stopping users from forgetting them to use it. We can always remind users during the code review process. Passing extra props for which a Provider is already in place seems wrong to me. In contributing guidelines we need to mention that one needs to externalize all strings.
